### PR TITLE
add expirasAt field and refactor SelfSAR check to SAR checks

### PR DESCRIPTION
**Pull request by alaypatel07**
This PR:
- add an `expiresAt` field to provide expiration time in UTC from #553 
- convert SelfSubjectAccessReview calls to SubjectAccessReview with
  username and scope from token review calls from MigToken
- refactors the code to add two getters on MigToken
- add few unexported status fields to keep the number of
     TokenReview calls low

### CODE REVIEW

File: pkg/compat/conversion/appsv1/zz_conversion_generated.go [hunk 1]
```diff
@@ -15,7 +15,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-// Code generated by conversion-gen-linux. DO NOT EDIT.
+// Code generated by conversion-gen-darwin. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply pkg/compat/conversion/appsv1/zz_conversion_generated.go:18]: bots expose my env dev env HAHA

File: pkg/apis/migration/v1alpha1/migtoken_types.go [hunk 2]
```diff
@@ -60,24 +69,43 @@ func init() {
 // If group is "*" then it means all API Groups
 // if namespace is "" then it means all cluster scoped resources
 func (r *MigToken) CanI(client k8sclient.Client, verb, group, resource, namespace, name string) (bool, error) {
-	sar := authapi.SelfSubjectAccessReview{
-		Spec: authapi.SelfSubjectAccessReviewSpec{
+	// sanity check if this token is validated
+	if r.Status.user == "" || r.Status.uid == "" {
```
Comment (dymurray) [inline pkg/apis/migration/v1alpha1/migtoken_types.go:73]: :+1:

File: pkg/apis/migration/v1alpha1/migtoken_types.go [hunk 3]
```diff
@@ -219,5 +230,95 @@ func (r *MigToken) GetProjectClient(client k8sclient.Client) (*projectv1.Project
...
+	if err != nil {
+		return err
+	}
+
+	var isOauthToken bool
+	switch {
+	case !tokenReview.Status.Authenticated:
+		// token is nolonger valid
+		return errors.New("invalid token")
+	// TODO: explore if the UI needs distinction between serviceaccount token
```
Comment (dymurray) [inline pkg/apis/migration/v1alpha1/migtoken_types.go:247]: :+1:

File: pkg/apis/migration/v1alpha1/migtoken_types.go [hunk 4]
```diff
@@ -219,5 +242,95 @@ func (r *MigToken) GetProjectClient(client k8sclient.Client) (*projectv1.Project
...
+	//   and oauth token, if so add a status field for it
+	case strings.Contains(tokenReview.Status.User.Username, "serviceaccount"):
```
Comment (dymurray) [inline pkg/apis/migration/v1alpha1/migtoken_types.go:261]: This is a smart way to do it. Cool

File: pkg/apis/migration/v1alpha1/migtoken_types.go [hunk 5]
```diff
@@ -219,5 +242,95 @@ func (r *MigToken) GetProjectClient(client k8sclient.Client) (*projectv1.Project
...
+		// service account token
+		isOauthToken = false
+	default:
+		isOauthToken = true
+	}
+
+	// TODO: explore adding r.Status.Authenticated field to reduce the
```
Comment (dymurray) [inline pkg/apis/migration/v1alpha1/migtoken_types.go:268]: :+1:

File: pkg/apis/migration/v1alpha1/migtoken_types.go [hunk 6]
```diff
@@ -219,5 +242,95 @@ func (r *MigToken) GetProjectClient(client k8sclient.Client) (*projectv1.Project
...
+		r.Status.scopes = scopes
+	}
+
+	// for oauth token set expiration date
+	if isOauthToken {
+		oauthAccessToken, err := r.GetOauthAccessToken(client)
+		if err != nil {
+			return err
+		}
+		t := metav1.NewTime(oauthAccessToken.GetObjectMeta().GetCreationTimestamp().
```
Comment (dymurray) [inline pkg/apis/migration/v1alpha1/migtoken_types.go:282]: :+1:

**Pull request merged**